torch/distributed/fsdp/common_utils.py

```diff
@@ -5,6 +5,37 @@
 from enum import auto, Enum
 
 
+class ShardingStrategy(Enum):
+    """
+    This specifies the sharding strategy to be used for distributed training by
+    :class:`FullyShardedDataParallel`.
+
+    - ``FULL_SHARD``: Parameters, gradients, and optimizer states are sharded.
+    For the parameters, this strategy unshards (via all-gather) before the
+    forward, reshards after the forward, unshards before the backward
+    computation, and reshards after the backward computation. For gradients, it
+    synchronizes and shards them (via reduce-scatter) after the backward
+    computation. The sharded optimizer states are updated locally per rank.
+    - ``SHARD_GRAD_OP``: Gradients and optimizer states are sharded during
+    computation, and additionally, parameters are sharded outside computation.
+    For the parameters, this strategy unshards before the forward, does not
+    reshard them after the forward, and only reshards them after the backward
```

Comment:
No need to change in this PR. This is prior to this PR. 

curious, why we still do unshard before forward? Is that because the peak memory consumption occurs during backward when computing gradients, and this mode will gradually re-shard params one unit at a time as the grad tensors are generated, so it still helps saving memory?

(for future) In some cases, will it help to have another mode that does not reshard at all, which would save two allgather?

Reply:
> curious, why we still do unshard before forward? Is that because the peak memory consumption occurs during backward when computing gradients, and this mode will gradually re-shard params one unit at a time as the grad tensors are generated, so it still helps saving memory?

Yes. Your understanding is correct.

> (for future) In some cases, will it help to have another mode that does not reshard at all, which would save two allgather?

If we have no all-gathers, then is that the same as `NO_SHARD` / DDP?